yoti-sdk-impl/src/main/java/com/yoti/api/client/spi/remote/call/factory/SignatureFactory.java

```diff
@@ -1,26 +1,42 @@
 package com.yoti.api.client.spi.remote.call.factory;
 
-import com.yoti.api.client.ProfileException;
-
 import java.security.GeneralSecurityException;
 import java.security.PrivateKey;
 import java.security.SecureRandom;
 import java.security.Signature;
 
+import static com.yoti.api.client.spi.remote.Base64.base64;
+
 public class SignatureFactory {
 
     private static final String BOUNCY_CASTLE_PROVIDER = "BC";
     private static final String SIGNATURE_ALGORITHM = "SHA256withRSA";
 
-    public byte[] create(byte[] message, PrivateKey privateKey) throws ProfileException {
-        try {
-            Signature signature = Signature.getInstance(SIGNATURE_ALGORITHM, BOUNCY_CASTLE_PROVIDER);
-            signature.initSign(privateKey, new SecureRandom());
-            signature.update(message);
-            return signature.sign();
-        } catch (GeneralSecurityException gse) {
-            throw new ProfileException("Cannot sign request", gse);
-        }
+    private final MessageFactory messageFactory;
+
+    public static SignatureFactory newInstance() {
+        return new SignatureFactory(new MessageFactory());
+    }
+
+    public SignatureFactory(MessageFactory messageFactory) {
+        this.messageFactory = messageFactory;
+    }
+
+    public String create(PrivateKey privateKey, String httpMethod, String path) throws GeneralSecurityException {
+        byte[] message = messageFactory.create(httpMethod, path, null);
+        return createSignature(message, privateKey);
+    }
+
+    public String create(PrivateKey privateKey, String httpMethod, String path, byte[] body) throws GeneralSecurityException {
+        byte[] message = messageFactory.create(httpMethod, path, body);
+        return createSignature(message, privateKey);
+    }
+
+    private String createSignature(byte[] message, PrivateKey privateKey) throws GeneralSecurityException {
+        Signature signature = Signature.getInstance(SIGNATURE_ALGORITHM, BOUNCY_CASTLE_PROVIDER);
+        signature.initSign(privateKey, new SecureRandom());
+        signature.update(message);
+        return base64(signature.sign());
```

Comment:
the method is called `createSignature` but actually, it creates a Signature obj and also signs it, generating a string that is the message signed. If the name of the class is SignatureFactory and we are thinking of having a different implementation of the `java.security.Signature` obj, it's a good idea to have it as a parameter during the init of the SignatureFactory class.

My suggestion is that we can return the Signature obj or if we want to sign also the message we can call the method in a different way, like `signMessage`. Also keep in mind that when we call `sign()` the state of the Signature object will be reverted to ` signature.initSign(privateKey, new SecureRandom());`. You can use the same instance for signing multiple times, there is no need to create a new one every time.

Reply:
> the method is called createSignature but actually, it creates a Signature obj and also signs it, generating a string that is the message signed.
Signing the signature is fine, that's just a part of our creation process.  The real problem then is returning a String rather than a Signature.  It takes byte[] message, so maybe it should be called signMessage.
> it's a good idea to have it as a parameter during the init of the SignatureFactory class
I doubt we'll need to use a different implementation of Signature.  That might be more than we need now.

> My suggestion is that we can return the Signature obj or if we want to sign also the message we can call the method in a different way, like signMessage
So the private method could be signMessage, and we could call the class SignedMessageFactory.  Would that be more appropriate?

> You can use the same instance for signing multiple times, there is no need to create a new one every time
If we do that, don't we need to synchronize on it?